internal/util/recoverable.go

```diff
@@ -0,0 +1,113 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ErrServiceStopped = fmt.Errorf("service stopped")
+)
+
+type Doable interface {
+	Do() error
+	Stop()
+}
+
+func NewRecoverableService(svc Doable, logger *log.Logger) *RecoverableService {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &RecoverableService{
+		service: svc,
+		stopped: make(chan stoppedService, 1),
+		log:     logger,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
+}
+
+type RecoverableService struct {
+	mu      sync.Mutex
+	running bool
+	service Doable
+	stopped chan stoppedService
+	log     *log.Logger
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func (m *RecoverableService) Start(_ context.Context) error {
```

Comment:
It looks like we never return an error from this function, only nil, so should we just change the signature to `func (m *RecoverableService) Start(_ context.Context)`? But it also looks like the context isn't used so we could drop it too?